stdpopsim/genomes.py

```diff
@@ -150,16 +204,162 @@ class Contig:
+            raise ValueError(
+                "Genomic elements are outside the genomic region"
+                " implied by the recombination map."
+            )
+        if not np.all(all_intervals[1:, 0] >= all_intervals[:-1, 1]):
+            raise ValueError("You cannot have overlapping genomic elements.")
+        return all_intervals
+
+    @property
+    def msp_mutation_rate_map(self):
+        """
+        breaks = [0,10,20] means [0,10) and [10,20)
+        """
+        if self.is_slim_incompatible:
+            return None
+        breaks = [0]
+        rates = []
+        for (start, end, t) in self.genomic_intervals:
+            if start not in breaks:
+                breaks.append(start)
+                rates.append(self.mutation_rate)
+            breaks.append(end)
+            rates.append((1 - self.slim_fractions[t]) * self.mutation_rate)
+        if not np.isclose(breaks[-1], int(self.length)):
+            breaks.append(int(self.length))
+            rates.append(self.mutation_rate)
+        assert len(breaks) == len(rates) + 1
+        return (breaks, rates)
+
+    @property
+    def slim_mutation_rate_map(self):
+        """
+        breaks = [10,20] means [0,10] and (10,20]
+        """
+        if self.is_slim_incompatible:
+            return None
+        # making sure the mut types are still concordant with the ge types
+        _check_mutation_types(self.mutation_types, self.genomic_element_types)
+        breaks, rates = self.msp_mutation_rate_map
+        assert breaks[0] == 0
+        slim_breaks = [b - 1 for b in breaks[1:]]
+        slim_rates = [self.mutation_rate - r for r in rates]
+        return (slim_breaks, slim_rates)
+
+    def add_mutation_type(self, mutation_type, genomic_element_type_id=0, proportion=0):
+        if genomic_element_type_id >= len(self.genomic_element_types):
+            raise ValueError(
+                f"Genomic element type of id {genomic_element_type_id} does not exist."
+            )
+        self.mutation_types.append(mutation_type)
+        self.genomic_element_types[genomic_element_type_id].mutation_type_ids.append(
+            len(self.mutation_types) - 1
+        )
+        self.genomic_element_types[genomic_element_type_id].proportions.append(
+            proportion
+        )
 
     def __str__(self):
         gmap = "None" if self.genetic_map is None else self.genetic_map.id
         s = (
             "Contig(length={:.2G}, recombination_rate={:.2G}, "
             "mutation_rate={:.2G}, genetic_map={})"
         ).format(
-            self.recombination_map.get_length(),
+            self.length,
             self.recombination_map.mean_recombination_rate,
             self.mutation_rate,
             gmap,
         )
         return s
+
+
+def _check_mutation_types(mutation_types, genomic_element_types):
+    mut_ids = [mtid for g in genomic_element_types for mtid in g.mutation_type_ids]
+    if set(range(len(mutation_types))) != set(mut_ids):
```

Comment:
they don't have to be the same, but the mut_ids in genomic element types must be a subset of all mutation types.